Userland/Applications/PixelPaint/Tools/TextTool.cpp

```diff
@@ -0,0 +1,225 @@
+{
+
+    if (m_is_typing) {
+        auto mouse_position = editor_stroke_position(event.layer_event().position(), 1);
+        m_mouse_is_over_text = m_ants_rect.contains(mouse_position);
+        m_editor->update_tool_cursor();
+    }
+
+    if (m_is_dragging) {
+        auto new_position = event.layer_event().position();
+        m_add_text_position = m_add_text_position + (new_position - m_drag_start_point);
+        m_drag_start_point = new_position;
+    }
+}
+void TextTool::on_mouseup(Layer*, MouseEvent&)
+{
+    m_is_dragging = false;
+}
+void TextTool::on_mousedown(Layer*, MouseEvent& event)
+{
+    auto start_text_region = [&] {
+        m_text_color = m_editor->color_for(event.layer_event());
+        m_is_typing = true;
+        m_text_editor->set_text(""sv);
+        m_add_text_position = event.layer_event().position();
+        m_editor->image().selection().begin_interactive_selection();
+        m_timer->start();
+        m_editor->update();
+    };
+
+    if (!m_is_typing) {
+        start_text_region();
+        return;
+    }
+
+    if (m_mouse_is_over_text) {
+        m_is_dragging = true;
+        m_drag_start_point = event.layer_event().position();
+    } else {
+        // User clicked somewhere outside the currently edited text region
+        // apply the current text and then start a new one where they clicked.
+        apply_text_to_layer();
+        reset_tool();
+        start_text_region();
+    }
+}
+
+GUI::Widget* TextTool::get_properties_widget()
+{
+    if (!m_properties_widget) {
+        m_properties_widget = GUI::Widget::construct();
+        m_properties_widget->set_layout<GUI::VerticalBoxLayout>();
+
+        auto& font_header = m_properties_widget->add<GUI::Label>("Current Font:");
+        font_header.set_text_alignment(Gfx::TextAlignment::CenterLeft);
+
+        m_font_label = m_properties_widget->add<GUI::Label>(m_selected_font->human_readable_name());
+
+        auto& change_font_button = m_properties_widget->add<GUI::Button>("Change Font...");
+        change_font_button.on_click = [&](auto) {
+            auto picker = GUI::FontPicker::construct(nullptr, m_selected_font, false);
+            if (picker->exec() == GUI::Dialog::ExecResult::OK) {
+                m_font_label->set_text(picker->font()->human_readable_name());
+                m_selected_font = picker->font();
+            }
+        };
+    }
+
+    return m_properties_widget.ptr();
+}
+
+void TextTool::on_second_paint(Layer const* layer, GUI::PaintEvent& event)
+{
+    if (!m_is_typing)
+        return;
+
+    GUI::Painter painter(*m_editor);
+    painter.add_clip_rect(event.rect());
+    painter.translate(editor_layer_location(*layer));
+    auto typed_text = m_text_editor->text();
+    auto text_width = max<int>(m_selected_font->width(typed_text), m_selected_font->width(" "sv));
+    auto text_height = m_selected_font->preferred_line_height() * max<int>((int)m_text_editor->line_count(), 1);
```

Comment:
Please use C++-style casts (`static_cast<int>`) instead of `(int)` everywhere

Reply:
Resolved in latest push :)